addons/sale/models/payment_transaction.py

```diff
@@ -32,19 +32,31 @@ def _compute_sale_order_ids_nbr(self):
             trans.sale_order_ids_nbr = len(trans.sale_order_ids)
 
     def _set_pending(self, state_message=None):
-        """ Override of payment to send the quotations automatically. """
-        super(PaymentTransaction, self)._set_pending(state_message=state_message)
+        """ Override of `payment` to send the quotations automatically.
 
-        for record in self:
-            sales_orders = record.sale_order_ids.filtered(lambda so: so.state in ['draft', 'sent'])
-            sales_orders.filtered(lambda so: so.state == 'draft').with_context(tracking_disable=True).write({'state': 'sent'})
-
-            if record.provider_id.code == 'custom':
-                for so in record.sale_order_ids:
-                    so.reference = record._compute_sale_order_reference(so)
-            # send order confirmation mail
+        :param str state_message: The reason for which the transaction is set in 'pending' state.
+        :return: updated transactions.
+        :rtype: `payment.transaction` recordset.
+        """
+        txs_to_process = super()._set_pending(state_message=state_message)
+
+        # When using webhooks, this function can be called multiple times (e.g. when the user comes
+        # back from a payment provider in a redirect flow and when the webhooks notify the db). To
+        # avoid duplicates, only process transactions for which the state has been updated.
+        for tx in txs_to_process:
```

Comment:
After the fix is applied, isn't it obvious that we don't want to do extra work for nothing? A simple and less dramatic comment should be enough.
```suggestion
        for tx in txs_to_process:  # Consider only transactions that are indeed set pending.
```